### DISCUSSION TITLE
[RFC FS-1072] - Support "without"ers for Anonymous Records

### DISCUSSION THREAD

**Discussion by cartermp** [fsharp/fslang-design#616] [Language and core library RFC discussions]
Discuss the [withouters RFC](https://github.com/fsharp/fslang-design/blob/main/RFCs/FS-1072-without-anonymous-records.md) here.

**Comment by cartermp** [comment]
### This is the syntax thread.

`with-` is kinda shitty.

`without` is a preferred keyword, but to support it would require making it a contextual keyword, which is something that F# doesn't do. And if we made `without` a contextual keyword, why not also make `with` a contextual keyword? It can be a slippery slope.

**Reply by baronfel** [reply]
> why not also make with a contextual keyword? 

If contextual keywords had been a thing when `with` was introduced, would it have been implemented in those terms? Essentially was a non-ideal implementation path chosen due to technical hurdles at the time? If so, then I wouldn't see `with` becoming a contextual keyword as a slippery slope so much as an alignment with the ideal.

**Reply by cartermp** [reply]
Perhaps, since there's nothing in inherently wrong with this aside from it being silly code:

```fsharp
let with = {| with = 12 |}
let b = {| with with Doot = 123 |}
```

If you can do this:

```fsharp
let without = {| doot="or"; without="you" |}
// val without: {| doot: string; without: string |}

let err = {| without without without |}
// val err: {| doot: string |}
```

then why should you also _not_ be able to do this?

```fsharp
let without = {| with="or"; without="you" |}
// val without: {| with: string; without: string |}

let with = {| without without without |}
// val with: {| with: string |}
```

**Reply by rmunn** [reply]
Other suggestions that were mentioned recently ("recently" in this case meaning since Jan 2021) in the https://github.com/fsharp/fslang-suggestions/issues/762 discussion:

In https://github.com/fsharp/fslang-suggestions/issues/762#issuecomment-800786340, @jcmrva suggested using a less-common antonym like `sans`.

In https://github.com/fsharp/fslang-suggestions/issues/762#issuecomment-890529764, @rmunn suggested using `not with`, which @jcmrva endorsed in https://github.com/fsharp/fslang-suggestions/issues/762#issuecomment-892706393.

The rest of the comments since Jan 2021 were along the lines of "I don't like `with-` and something else would be better". The only syntaxes anyone has proposed seem to be the original `without` proposal, the `with-` syntax that got several negative responses, `not with`, and `sans`.

**Reply by rmunn** [reply]
Now to offer my opinion: I don't actually like `not with` now that I think about it, for two reasons:

1) How would the parser handle `let foo = {| bar not with x; with y = 5 |}`? It seems to me that implementing `not with` would also require implementing contextual keywords in some form to be able to tell the difference between `not with` and `with` in that assignment.
2) The following legal (but extremely silly) code would either become illegal, or would be a nasty corner case for the parser:

```fs
let not = {| x = 5; y = 8 |}
let foo = {| not with y = 13 |}
printfn "%A" foo
// Should print x = 5; y = 13
```

While that's not much different from the `without without without` example\* in terms of parser difficulty, at least `without` is a single word, as opposed to a keyword pair. A single word is probably (I assume, not having looked at parser internals) a little easier to cope with.

\* Which is rapidly approaching "Buffalo buffalo Buffalo buffalo buffalo buffalo Buffalo buffalo" territory. (If you've never see that sentence before, it's a totally valid sentence according to the rules of English grammar. The challenge is to make sense of it. Hint (rot13): "gb ohssnyb" pna or n ireo zrnavat "gb ohyyl").

**Reply by hvester** [reply]
If using `without` is out of the question how about using `-` in front of the label to be dropped as below?

```fsharp
let data = {| X = 1; Y = "abc"; Z = 6 |}
// val data: {| X: int; Y: string; Z: int |}

let subset = {| data with -Z |}
// val subset: {| X: int; Y: string |}
```

This would allow adding and dropping labels within one `with`:

```fsharp
let data = {| X = 1; Y = "abc"; Z = 6; W = 0 |}
// val data: {| W: int; X: int; Y: string; Z: int |}

let subset = {| data with A = 3; -Z; -W |}
// val subset: {| A: int; X: int; Y: string |}
```
I would prefer this to both `with-`and `not with` and would assume that is is more understandable than `with-` for someone seeing the syntax for the first time.

**Reply by uxsoft** [reply]
> Perhaps, since there's nothing in inherently wrong with this aside from it being silly code:
> 
> ```fsharp
> let with = {| with = 12 |}
> let b = {| with with Doot = 123 |}
> ```
> 
> If you can do this:
> 
> ```fsharp
> let without = {| doot="or"; without="you" |}
> // val without: {| doot: string; without: string |}
> 
> let err = {| without without without |}
> // val err: {| doot: string |}
> ```
> 
> then why should you also _not_ be able to do this?
> 
> ```fsharp
> let without = {| with="or"; without="you" |}
> // val without: {| with: string; without: string |}
> 
> let with = {| without without without |}
> // val with: {| with: string |}
> ```

The IDE should color everything to show if it's a keyword, a property, or a variable. So, even in this silly code, all should be clear to the user?

**Comment by dsyme** [comment]
I'm ok with `without` if it can be made to work as a contextual keyword, not making it a full keyword

**Comment by uxsoft** [comment]
I'm very much not a fan of `with-`. This mixing of symbols and letters is weird for a keyword, also what happens when people format it as in the English language and put space inbetween (`with -`)? I reckon the compiler will get confused and will reply with some unhelpful error message that - wasn't expected which will confuse people even more.

I like @baronfel's suggestion to make both with and without a contextual keyword and implement this in a clean and consistent way. If we don't we'll be stuck with this forever as it's much harder to take things out without breaking people than to put them in.

**Comment by purkhusid** [comment]
Would it be possible to derive an anonymous record from a named record?
E.g. if I have this record
```
type Donk = { A: int; B: string; C: double }
```
Then I would like to be able to do something like:
```
let originalDonk = { A = 1; B = "yo"; C = 2.0 }
let anonymousDerivedDonk = { originalDonk without C }
```

And now anonymousDerivedDonk would be of the type
```
{| A: int; B: string |}
```

**Reply by Happypig375** [reply]
Existing code:
```fs
type Donk = { A: int; B: string; C: double }
let originalDonk = { A = 1; B = "yo"; C = 2.0 }
let anonymousDerivedDonk = {| originalDonk with D = 2 |}
```
I would think that `{| originalDonk without C |}` is required instead of `{ originalDonk without C }`.

**Comment by purkhusid** [comment]
Has there been any progress on this feature? This would be a pretty powerful addition to F#. Especially for data pipeline kind of tasks.

**Reply by T-Gro** [reply]
There is a now abandoned prototype, which could be picked up if there is anyone interested into pursuing this.
https://github.com/dotnet/fsharp/pull/11911

**Comment by brianrourkeboll** [comment]
## Circumfix: `...expr without member`, `...expr without ...expr`

One way that https://github.com/fsharp/fslang-suggestions/issues/1253 could help make `without` work in a backwards-compatible way would be for it to be valid only when used together with a spread, since the spread operator itself would be newly-valid syntax. A `without` token directly following a spread expression could never have been valid before, since the spread expression itself wasn't valid.

Examples:

```fsharp
let r = {| A = 1; B = 2; C = 3 |}
let s = {| ...r without B |}
```

```fsharp
type R = { A : int; B : int; C : int }
type S = { ...R without B }
```

We would probably want `without` to work for _sets_ of members, too.

Given

```fsharp
type R = { A : int; B : int; C : int; D : int }
type S = { B : int; C : int }
type T = { ...R without B }
```

I think we would probably want

```fsharp
type U = { ...R without ...S }
```

since using plain `without S` could be confused for excluding a member `S` if it also existed.

Likewise:

```fsharp
let r = {| A = 1; B = 2; C = 3; D = 4 |}
let s = {| B = 2; C = 3 |}
let t = {| ...r without ...s |}
```

## Set difference notation: `...expr \ member`, `...expr \ ...expr`

Alternatively, we could use a symbol like `\`, _à la_ its usage for [set difference](https://en.wikipedia.org/wiki/Complement_(set_theory)#Relative_complement).

`\` isn't currently valid syntax on its own, so using it for this wouldn't be a breaking change.

So, to repeat the examples above with `\` in place of `without`:

```fsharp
let r = {| A = 1; B = 2; C = 3 |}
let s = {| ...r \ B |}
```

```fsharp
type R = { A : int; B : int; C : int }
type S = { ...R \ B }
```

```fsharp
let r = {| A = 1; B = 2; C = 3; D = 4 |}
let s = {| B = 2; C = 3 |}
let t = {| ...r \ ...s |}
```

```fsharp
type R = { A : int; B : int; C : int; D : int }
type S = { B : int; C : int }
type T = { ...R \ B }
type U = { ...R \ ...S }
```

But maybe `\` is too exotic or too subtle.

**Reply by T-Gro** [reply]
I like how this composes.
`...` is used to convert either a type or a value to a `spreadable` value/type.
`;` would be typically used to concatenate spreadables, `\` would be subtracting them.

So first, if you want to change types, you convert them to spreadables.
And then you have an algebra of operations on spredables.
`;` - take symbols from LHS and RHS spreadables and concat
`\` - subtract RHS symbols from LHS
Maybe even anon unions could be created as an operation on spreadables? ( creating an anonymous union, a choice between LHS fields and RHS fields)

**Reply by vzarytovskii** [reply]
I personally like the keyworded version better, since it doesn't introduce obscure "custom operator like" syntax (\\).

Using the spread operator is great though. The main reason I didn't finish the without feature is that there was no agreement whether it should go together with spread feature or separately.

**Reply by smoothdeveloper** [reply]
would we prefer to distinguish the `without` that takes a type, and the `without` that takes a value, this way:

```fsharp
{ a without ...<MyType> }
{ a without ...b }
```

I'm concerned that not distinguishing both would make it less obvious to distinguish the bare field(s) remover: `{ be without manner; burden; past; future; }`.

**Reply by brianrourkeboll** [reply]
I personally think the `without ...SomeTy` (which would only work with a record type or record value after the `...`) would be enough of a differentiator from `without SomeField`. I think angle brackets would be more confusing, especially since they seem to imply some kind of generic-ness which would not be the case here.

**Comment by brianrourkeboll** [comment]
## "Rest" patterns + field patterns

I did not think of it till now, but some form of "rest" patterns combined with record field patterns could offer another way to subsume `without` at the value level:

```fsharp
type R = { A : int; B : int; C : int }

let r = { A = 1; B = 2; C = 3 }
let { A = _; ...rest } = r // rest = {| B = 2; C = 3 |}
```

**Reply by T-Gro** [reply]
Interesting.
I am personally not sure if record patterns are the right angle of growing the language further (not even a big fan/user of existing record patterns), but curious what others think.

To me it feels like reading the intention backwards (having to use pattern matching syntax in order to construct a value).
As opposed to following `I want 'r' without 'A'` as close as possible.
